lib/fluent/plugin/out_s3.rb

```diff
@@ -80,6 +82,15 @@ def start
       else
         options[:credential_provider] = AWS::Core::CredentialProviders::EC2Provider.new({:retries => @aws_iam_retries})
       end
+
+      # If a role ARN was provided, use the EC2 credential provider to connect to STS.
+      # Then use an AssumeRole provider when connecting to S3.
+      if @sts_role_arn
+        sts_conn = AWS::STS.new(options)
+        provider = AWS::Core::CredentialProviders::AssumeRoleProvider.new({:sts => sts_conn, :role_arn => @sts_role_arn, :role_session_name => @sts_role_session_name})
+        options[:credential_provider] = provider
```

Comment:
To avoid ambiguity, I changed this to explicitly delete the access and secret keys from the options hash after establishing the STS connection.

Environment variables are handled by another credential_provider, which we are simply replacing in the options hash before attempting the connection to S3.